json_object.c

```diff
@@ -1775,6 +1875,89 @@ static int json_object_deep_copy_recursive(struct json_object *src, struct json_
 	return 0;
 }
 
+/*
+ * The actual guts of json_object_deep_copy(), with a few additional args
+ * needed so we can keep track of where we are within the object tree.
+ *
+ * Note: caller is responsible for freeing *dst if this fails and returns -1.
+ */
+static int json_object_deep_copy_recursive_len(struct json_object *src, struct json_object *parent,
```

Comment:
json_object_deep_copy_recursive_len() looks like a lot of code duplication just to pass a different shallow_copy function pointer.  Instead, add another argument to the existing json_object_deep_copy_recursive() and have it use whichever one is non-NULL.